src/ucp/rndv/proto_rndv.c

```diff
@@ -631,15 +637,180 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+
+    *rkey_length = *(const uint32_t*)p;
+    p            = UCS_PTR_BYTE_OFFSET(p, sizeof(uint32_t));
+
+    *rkey_buffer = p;
+    p            = UCS_PTR_BYTE_OFFSET(p, *rkey_length);
+
+    return p;
+}
+
+static ucs_status_t
+ucp_proto_rndv_unpack_iov_rkeys(const void *buffer, size_t length, ucp_ep_h ep,
+                                ucp_ep_config_t *ep_config, ucp_request_t *req)
+{
+    const void *p = buffer;
+#if UCS_ENABLE_ASSERT
+    const void *p_end = UCS_PTR_BYTE_OFFSET(p, length);
+#endif
+    uint32_t iov_count, iov_index;
+    ucp_rkey_h rkey;
+    const void *rkey_buffer;
+    size_t rkey_length;
+    uint64_t remote_address;
+    size_t remote_size, accumulate_size;
+    ucs_status_t status;
+
+    p = ucp_proto_rndv_unpack_iov_count(p, &iov_count);
+    if (iov_count == 0) {
+        return UCS_ERR_NO_MESSAGE;
+    }
+
+    req->send.rndv.rma_array = ucs_malloc(
+            iov_count * sizeof(*req->send.rndv.rma_array), "proto_rndv");
+    if (req->send.rndv.rma_array == NULL) {
+        ucs_error("failed to allocate rma array");
+        status = UCS_ERR_NO_MEMORY;
+        goto finish;
+    }
+
+    accumulate_size = 0;
+    for (iov_index = 0; iov_index != iov_count; ++iov_index) {
+        p = ucp_proto_rndv_unpack_iov_rkey_buffer(p, &rkey_buffer, &rkey_length,
+                                                  &remote_address,
+                                                  &remote_size);
+        ucp_proto_rndv_check_rkey_length(remote_address, remote_size, "rts");
+        ucp_proto_rndv_check_rkey_length((uint64_t)rkey_buffer, rkey_length,
+                                         "rts");
+
+        status = ucp_ep_rkey_unpack_internal(
+                ep, rkey_buffer, rkey_length, ep_config->key.reachable_md_map,
+                ep_config->rndv.proto_rndv_rkey_skip_mds, &rkey);
+        if (status != UCS_OK) {
+            goto destroy_rkey;
+        }
+        accumulate_size                                    += remote_size;
+        req->send.rndv.rma_array[iov_index].accumulate_size = accumulate_size;
+        req->send.rndv.rma_array[iov_index].remote_address  = remote_address;
+        req->send.rndv.rma_array[iov_index].remote_size     = remote_size;
+        req->send.rndv.rma_array[iov_index].rkey            = rkey;
+        ucs_debug("index=%u rkey_size=%lu address=%p size=%lu", iov_index,
+                  rkey_length, (void*)remote_address, remote_size);
+    }
+    req->send.rndv.rma_count = iov_count;
+    req->send.rndv.rma_index = 0;
+    status                   = UCS_OK;
+    goto finish;
+
+destroy_rkey:
+    req->send.rndv.rma_count = iov_index;
+    ucp_proto_rndv_iov_rkeys_destroy(req);
+finish:
+    ucs_assertv(p <= p_end, "bounds error 0x%" PRIx64 ">0x%" PRIx64,
+                (uint64_t)p, (uint64_t)p_end);
+    return status;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_proto_rndv_unpack_common_rkey(
+        const void *buffer, size_t length, uint64_t remote_address,
```

Comment:
I would prefer to keep consistent with `ucp_proto_rndv_receive_start` names
```suggestion
        const void *rkey_buffer, size_t rkey_length, uint64_t remote_address,
```

Reply:
done